src/hexen/g_game.c

```diff
@@ -407,22 +407,49 @@ void G_BuildTiccmd(ticcmd_t *cmd, int maketic)
     {
         gamekeydown[key_arti_all] = false;     // Use one of each artifact
         cmd->arti = NUMARTIFACTS;
+    }
+	else if (gamekeydown[key_arti_invulnerability] && !cmd->arti
+             && !players[consoleplayer].powers[pw_invulnerability])
+    {
+        gamekeydown[key_arti_invulnerability] = false;
+        cmd->arti = arti_invulnerability;
     }
     else if (gamekeydown[key_arti_health] && !cmd->arti
              && (players[consoleplayer].mo->health < MAXHEALTH))
     {
         gamekeydown[key_arti_health] = false;
         cmd->arti = arti_health;
+	}
+	else if (gamekeydown[key_arti_superhealth] && !cmd->arti
+             && (players[consoleplayer].mo->health < MAXHEALTH))
+    {
+        gamekeydown[key_arti_superhealth] = false;
+        cmd->arti = arti_superhealth;
     }
-    else if (gamekeydown[key_arti_poisonbag] && !cmd->arti)
+	else if (gamekeydown[key_arti_healingradius] && !cmd->arti)
+	{
+		gamekeydown[key_arti_healingradius] = false;
+		cmd->arti = arti_healingradius;
+	}
+	else if (gamekeydown[key_arti_torch] && !cmd->arti)
     {
-        gamekeydown[key_arti_poisonbag] = false;
-        cmd->arti = arti_poisonbag;
+        gamekeydown[key_arti_torch] = false;
+        cmd->arti = arti_torch;
     }
-    else if (gamekeydown[key_arti_blastradius] && !cmd->arti)
+    else if (gamekeydown[key_arti_egg] && !cmd->arti)
```

Comment:
This code is becoming very repetitive. I think you could restructure it to become more table-driven, eg.

```
struct {
  int *key;
  int artifact;
} artifact_keys[] = {
  { &key_arti_egg, arti_egg},
  { &key_arti_fly, arti_fly},
  ... etc
```

You'll need to keep some separate checks for eg. the superhealth artifact though since it has additional logic.